docs/ingestion-sink/index.md

```diff
@@ -0,0 +1,224 @@
+
+### Encoding
+
+When writing messages to Google Cloud Storage or BigQuery, the message received
+from Google Cloud Pub/Sub will be encoded as a JSON object.
+
+When `OUTPUT_FORMAT` is unspecified or `raw` messages will have bytes encoded as
+a `"payload"` field with base64 encoding, and each attribute encoded as field.
+This is the format used for `payload_bytes_raw.*` tables.
+
+When `OUTPUT_FORMAT` is `decoded` messages will have bytes encoded as with
+`OUTPUT_FORMAT=raw`, but attributes will be encoded using the nested metadata
+format of decoded pings. This is the format used for `payload_bytes_decoded.*`
+tables.
+
+When `OUTPUT_FORMAT` is `payload` messages will have bytes decoded as JSON, and
+transformed to coerce types and use snake case for compatibility with BigQuery.
+This is the format used for `*_live.*` tables. This requires specifying a local
+path to a gzipped tar archive that contains BigQuery table schemas as
+`SCHEMAS_LOCATION`. If messages bytes are compressed then
+`INPUT_COMPRESSION=gzip` must also be specified to ensure they are decompressed
+before they are decoded as JSON.
+
+When `OUTPUT_FORMAT` is `beam` messages will have bytes encoded as with
+`OUTPUT_FORMAT=raw`, but attributes will be encoded as an `"attributeMap"` field
+that contains a JSON object. This is the same format as produced by ingestion-beam
+ when using `--outputType=file` and `--outputFileFormat=json`.
+
+### Google Cloud Storage file prefix
+
+Google Cloud Storage files are named like:
+
+    $OUTPUT_BUCKET/{UUID.randomUUID().toString()}.ndjson
+
+or if `OUTPUT_TABLE` and `BIG_QUERY_OUTPUT_MODE` are specified:
+
+    $OUTPUT_BUCKET/OUTPUT_TABLE=$OUTPUT_TABLE/{UUID.randomUUID().toString()}.ndjson
+
+for example, with `OUTPUT_BUCKET=gs://test-bucket/test-output`:
+
+    gs://test-bucket/test-output/ad715b24-7500-45e2-9691-cb91e3b9c2cc.ndjson
+
+or with `OUTPUT_BUCKET=gs://test-bucket/test-output`, `OUTPUT_TABLE=my_dataset.raw_table`, and
+`BIG_QUERY_OUTPUT_MODE=file_loads`:
+
+    gs://test-bucket/test-output/OUTPUT_TABLE=my_dataset.raw_table/3b17c648-f8b9-4250-bdc1-5c2e472fdc26.ndjson
+
+## Executing
+
+### Locally with Docker
+
+The provided `bin/mvn` script downloads and runs maven via docker so that less
+setup is needed on the local machine. For prolonged development performance is
+likely to be significantly better, especially in MacOS, if `mvn` is installed and
+run natively without docker.
+
+```bash
+# create a test input file
+mkdir -p tmp/
+echo '{"payload":"dGVzdA==","attributeMap":{"host":"test"}}' > tmp/input.ndjson
+
+# consume messages from the test file, decode and re-encode them, and write to a directory
+PASS_ENV="INPUT_PIPE=tmp/input.ndjson OUTPUT_PIPE=tmp/output.ndjson" ./bin/mvn compile exec:java
+
+# check that the message was delivered
+cat tmp/output.ndjson
+
+# read message from stdin and write to stdout
+echo tmp/input.ndjson | PASS_ENV="INPUT_PIPE=- OUTPUT_PIPE=-" ./bin/mvn compile exec:java
```

Comment:
```suggestion
cat tmp/input.ndjson | PASS_ENV="INPUT_PIPE=- OUTPUT_PIPE=-" ./bin/mvn compile exec:java

# read message from stdin and write to gcs
# note that $ needs to be escaped with \ to prevent shell substitution
cat tmp/input.ndjson | PASS_ENV="\
  INPUT_PIPE=- \
  OUTPUT_BUCKET=gs://my_bucket/\${document_type:-UNSPECIFIED}/ \
" ./bin/mvn compile exec:java
```